File: ruby/beer_song.rb
class BeerSong
  attr_reader :bottles, :nouns

  def initialize
    @bottles = Hash.new{|h,k| h[k]= k.to_s + " bottles"}.merge({1=>"1 bottle", 0=>"no more bottles", -1=>"99 bottles"})
    @nouns = Hash.new("one").merge({1=>"it"})
  end

  def verse(num)
    first_line(num) + end_of_verse(num)
  end

  def verses(start, last, lyrics=[])
    until start<last
      lyrics<<verse(start)
      start-=1
    end
    lyrics.join("\n")
  end

  def lyrics
    verses(99, 0)
  end

  def first_line(num)
    "#{@bottles[num].capitalize} of beer on the wall, #{@bottles[num]} of beer.\n"
  end

  def end_of_verse(num)
    final_verse(num) + "#{@bottles[num-1]} of beer on the wall.\n"
  end

  def final_verse(num)
    num == 0 ? "Go to the store and buy some more, ": "Take #{@nouns[num]} down and pass it around, "
  end
end

File: js/beer-song.js
BeerSong = function() {};

BeerSong.prototype.verse = function (num, str="") {

  function numBottles(num) {
    if (num == -1){return "99 bottles of beer"}
    else if (num == 0){return "No more bottles of beer"}
    else if (num == 1){return "1 bottle of beer"}
    else {return num + " bottles of beer"}
  };

  function firstLine(bottles) {
    return bottles + " on the wall, " + bottles.toLowerCase() + ".\n"
  };

  function lastLine(num, str="") {
    if (num == 0) {str += "Go to the store and buy some more, "}
    else if (num == 1) {str += "Take it down and pass it around, "}
    else {str += "Take one down and pass it around, "};
    return str + numBottles(num-1).toLowerCase()
  };

  return firstLine(numBottles(num)) + lastLine(num) + " on the wall.\n"
};

BeerSong.prototype.sing = function (start, stop=0, string="") {
  for (i=start; i>=stop; i--) {
    string+=new BeerSong().verse(i)
    if (i != stop){string+="\n"};
  };
  return string
};

module.exports = BeerSong;
